orderbook/usecase/orderbook_usecase_test.go

```diff
@@ -0,0 +1,314 @@
+package orderbookusecase_test
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+
+	cltypes "github.com/osmosis-labs/osmosis/v25/x/concentrated-liquidity/types"
+	"github.com/osmosis-labs/sqs/domain/mocks"
+	orderbookdomain "github.com/osmosis-labs/sqs/domain/orderbook"
+	orderbookusecase "github.com/osmosis-labs/sqs/orderbook/usecase"
+	"github.com/osmosis-labs/sqs/sqsdomain/cosmwasmpool"
+
+	"github.com/osmosis-labs/osmosis/osmomath"
+	"github.com/osmosis-labs/osmosis/v25/app/apptesting"
+)
+
+type OrderbookUsecaseTestSuite struct {
+	apptesting.ConcentratedKeeperTestHelper
+}
+
+func TestOrderbookUsecaseTestSuite(t *testing.T) {
+	suite.Run(t, new(OrderbookUsecaseTestSuite))
+}
+
+func (s *OrderbookUsecaseTestSuite) TestCreateFormattedLimitOrder() {
+	testCases := []struct {
+		name          string
+		order         orderbookdomain.Order
+		quoteAsset    orderbookdomain.Asset
+		baseAsset     orderbookdomain.Asset
+		setupMocks    func(orderbookRepo *mocks.OrderbookRepositoryMock, tokensUsecase *mocks.TokensUsecaseMock)
+		expectedError string
+		expectedOrder orderbookdomain.LimitOrder
+	}{
+		{
+			name: "tick not found",
+			order: orderbookdomain.Order{
+				TickId: 99, // Non-existent tick ID
+			},
+			expectedError: "tick not found",
+			setupMocks: func(orderbookRepo *mocks.OrderbookRepositoryMock, tokensUsecase *mocks.TokensUsecaseMock) {
+				orderbookRepo.GetTickByIDFunc = func(poolID uint64, tickID int64) (orderbookdomain.OrderbookTick, bool) {
+					return orderbookdomain.OrderbookTick{
+						Tick: &cosmwasmpool.OrderbookTick{
+							TickId: tickID,
+						},
+					}, false
+				}
+			},
+		},
+		{
+			name: "error parsing quantity",
+			order: orderbookdomain.Order{
+				Quantity: "invalid", // Invalid quantity
+			},
+			expectedError: "error parsing quantity",
+			setupMocks: func(orderbookRepo *mocks.OrderbookRepositoryMock, tokensUsecase *mocks.TokensUsecaseMock) {
+				orderbookRepo.GetTickByIDFunc = func(poolID uint64, tickID int64) (orderbookdomain.OrderbookTick, bool) {
+					return orderbookdomain.OrderbookTick{
+						Tick: &cosmwasmpool.OrderbookTick{
+							TickId: tickID,
+						},
+					}, true
+				}
+			},
+		},
+		{
+			name: "error parsing placed quantity",
```

Comment:
We do [return](https://github.com/osmosis-labs/sqs/blob/e4ce75541f31c4de7baf6da3e8b95fae6744f03e/orderbook/usecase/orderbook_usecase.go#L383) quantity as `int64` for the clients, if for example `quantity` overflows we will catch it with `strconv.ParseInt`. At this point I am trying to have exactly same precision logic as FE had. If we want to have some adjusments regarding that, maybe we can create a separate task to track that?

Reply:
My point is that it will frequently overflow in `strconv.ParseInt`, making the system unusable.

Consider 1000 DAI which is 1000 * 10**18 in base units.

```
>>> 2**63 - 1 - 1000 * 10**18
-990776627963145224193
```

Just $1000 of DAI overflows here so we should use the appropriate chain types to prevent that.

Reply:
Ah now I see you point, thanks elaboration! :pray: The idea is that currently we support only stables ( `USDC`, `USDT` ) as a quotes each of which has `6` decimal places,  meaning that paid amount should be larger than `1*10**12` in order to overflow:

```
>>> 2**63 - 1 - 1*10**12 * 10**6
```

In order to support non stables as a quotes would require some adjustments on FE side to, meaning we can be reasurred that somebody from product will not enable support of more asset types without anyone noticing from engineering team ( including us ) in advance.

Having this information, do you think we still should go and implement this change within scope of this task?

Reply:
I'm not sure I understand how stables as quotes excludes the possibility of overflow.

An order can be placed in both base and quote denominations so amounts are not limited to only the listed stables precision. 

Additionally, there are 18 decimal stables like DAI which we can trivially add support for.

I would vote to do it right now so that we don't need to worry about it. I also have low confidence in our cross-team communication on breaking changes so being extra diligent around such edge cases would pay-off down the road, helping to avoid incidents.

Reply:
Thanks! I have implemented your suggested improvement [here](https://github.com/osmosis-labs/sqs/pull/492/files#diff-ae675fb65861261e203f8bf1a1aacfd68815f3cb763d28db817797e9a88904b3R278) and have adjusted tests to check cases for possible `osmomath.Dec` overflows. Would that work?